launcher/src/main/java/org/apache/spark/launcher/SparkLauncher.java

```diff
@@ -0,0 +1,371 @@
+  public SparkLauncher setDeployMode(String mode) {
+    checkNotNull(mode, "mode");
+    this.deployMode = mode;
+    return this;
+  }
+
+  /**
+   * Set the main application resource. This should be the location of a jar file for Scala/Java
+   * applications, or a python script for PySpark applications.
+   */
+  public SparkLauncher setAppResource(String path) {
+    checkNotNull(path, "path");
+    this.userResource = path;
+    return this;
+  }
+
+  /** Sets the application class name for Java/Scala applications. */
+  public SparkLauncher setClass(String userClass) {
+    checkNotNull(userClass, "userClass");
+    this.userClass = userClass;
+    return this;
+  }
+
+  /** Adds command line arguments for the application. */
+  public SparkLauncher addArgs(String... args) {
+    for (String arg : args) {
+      checkNotNull(arg, "arg");
+      userArgs.add(arg);
+    }
+    return this;
+  }
+
+  /** Adds a jar file to be submitted with the application. */
+  public SparkLauncher addJar(String path) {
+    checkNotNull(path, "path");
+    jars.add(path);
+    return this;
+  }
+
+  /** Adds a file to be submitted with the application. */
+  public SparkLauncher addFile(String path) {
+    checkNotNull(path, "path");
+    files.add(path);
+    return this;
+  }
+
+  /** Adds a a python file / zip / egg to be submitted with the application. */
+  public SparkLauncher addPyFile(String path) {
+    checkNotNull(path, "path");
+    pyFiles.add(path);
+    return this;
+  }
+
+  /** Enables verbose reporting for SparkSubmit. */
+  public SparkLauncher setVerbose(boolean verbose) {
+    this.verbose = verbose;
+    return this;
+  }
+
+  /**
+   * Starts a new thread that will run the Spark application.
+   * <p/>
+   * The application will run on a separate thread and use a separate, isolated class loader.
+   * No classes or resources from the current thread's class loader will be visible to the app.
+   * <p/>
+   * This mode does not support certain configuration parameters, like configuring the amount of
+   * driver memory or custom driver command line options. If such configuration is detected, an
+   * exception will be thrown.
+   * <p/>
+   * This is extremely experimental and should not be used in production environments.
+   * <p/>
+   * NOTE: SparkSubmit uses system properties to propagate some configuration value to the app
+   * are run concurrently, they may affect each other's configurations.
+   * <p/>
+   * NOTE: for users running JDK versions older than 8, this option can add a lot of overhead
+   * to the VM's perm gen.
+   *
+   * @param exceptionHandler Optional handler for handling exceptions in the app thread.
+   * @param daemon Whether to start a daemon thread.
+   * @return A non-daemon thread that will run the application using SparkSubmit. The thread will
+   *         already be started.
+   */
+  public Thread start(Thread.UncaughtExceptionHandler handler, boolean daemon) throws IOException {
```

Comment:
Hm, can you motivate the use case of this? Is there a scenario that a user would prefer to launch it in a thread instead? Also, if we do decide to support this we should probably name this something more specific, since it's confusing to have a `start` and a `launch` and have them do different things.